### PR TITLE
Wasm/WASI target implementation

### PR DESCRIPTION
**Pull request by igoriakovlev**
This is Wasm/WASI target implementation. It based on native version with hardcoded tzdb.

1. Extracted native implementation into intermediate sourceset.
2. Added WASI implementation
3. Added timezones-full database project

Before building timezones project you should run generateWasmWasiZoneInfo task to download and convert timezones db (nodejs and npm should be installed in the system). We decided to put the result of this task into the repository but for review it is not generated yet.

### CODE REVIEW

File: timezones/full/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    @OptIn(ExperimentalWasmDsl::class)
+    wasmWasi {
+        nodejs()
+        NpmResolverPlugin.apply(project) //Workaround KT-66373
+    }
+
+    sourceSets {
+        commonMain {
+            dependencies {
+                compileOnly(project(":kotlinx-datetime"))
```
Comment (igoriakovlev) [inline timezones/full/build.gradle.kts:34]: Applying timezones without datetime lead to broken compilation (or exception on usage times zone when PL is enabled). This helps us to avoid version collisions on gradle side and circullar dependency while testing.

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    check(timeZonesProvider != provider) { "TimeZone database redeclaration" }
+    timeZonesProvider = provider
+}
+
+@InternalDateTimeApi
+private var timeZonesProvider: TimeZonesProvider? = null
+
+@OptIn(InternalDateTimeApi::class)
+private fun zoneDataByName(name: String): ByteArray =
+    timeZonesProvider?.zoneDataByName(name) ?: error("TimeZones are not supported")
```
Comment (igoriakovlev) [inline core/wasmWasi/src/internal/TimeZonesInitializer.kt:34]: We need to decide how to lead user to use timezones dependency in this case

**Comment by igoriakovlev** [comment]
The build fails because of none of TZ binaries in the PR.

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+package kotlinx.datetime.internal
+
+@RequiresOptIn
+internal annotation class InternalDateTimeApi
+
+/*
+This is internal API which is not intended to use on user-side.
+ */
+@InternalDateTimeApi
+public interface TimeZonesProvider {
```
Comment (dkhalanskyjb) [inline core/wasmWasi/src/internal/TimeZonesInitializer.kt:15]: If you mark this as `internal` and `@Suppress` the warning, will anything break? It would be nicer not to expose unnecessary symbols to autocompletion.

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+package kotlinx.datetime.internal
+
+@RequiresOptIn
+internal annotation class InternalDateTimeApi
+
+/*
+This is internal API which is not intended to use on user-side.
+ */
```
Comment (dkhalanskyjb) [inline core/wasmWasi/src/internal/TimeZonesInitializer.kt:13]: I think Dokka will generate the HTML for this API entry if we don't `/** @suppress */` it, no?

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+@InternalDateTimeApi
+private var timeZonesProvider: TimeZonesProvider? = null
+
+@OptIn(InternalDateTimeApi::class)
+private fun zoneDataByName(name: String): ByteArray =
+    timeZonesProvider?.zoneDataByName(name) ?: error("TimeZones are not supported")
+
+@OptIn(InternalDateTimeApi::class)
+private fun getTimeZones(): Set<String> =
+    timeZonesProvider?.getTimeZones() ?: emptySet()
```
Comment (dkhalanskyjb) [inline core/wasmWasi/src/internal/TimeZonesInitializer.kt:38]: This should contain at least `UTC`.

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+internal class TzdbOnData: TimeZoneDatabase {
+    override fun rulesForId(id: String): TimeZoneRules =
+        readTzFile(zoneDataByName(id)).toTimeZoneRules()
+
+    override fun availableTimeZoneIds(): Set<String> = getTimeZones()
+}
```
Comment (dkhalanskyjb) [inline core/wasmWasi/src/internal/TimeZonesInitializer.kt:45]: (Minor) Is there a reason to have the private functions above instead of putting the implementation here directly?

File: timezones/full/build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+val tzdbDirectory = File(project.projectDir, "tzdb")
+
+kotlin {
+    @OptIn(ExperimentalWasmDsl::class)
+    wasmWasi {
+        nodejs()
+        NpmResolverPlugin.apply(project) //Workaround KT-66373
+    }
+
+    sourceSets {
```
Comment (dkhalanskyjb) [inline timezones/full/build.gradle.kts:31]: Please also use the `sourceSets.all` block from the other build scripts in the library to set the directory structure like `common/test`.

File: timezones/full/build.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+val tzdbDownloadAndCompile = task<TzdbDownloadAndCompile>("tzdbDownloadAndCompile") {
+    dependsOn(timeTzdbInstall)
+    timeTzdbDirectory.set(timeTzdbInstall.outputDirectory)
+    outputDirectory.set(tzdbDirectory)
+    ianaVersion.set(tzdbVersion)
+    convertType.set(ConvertType.LARGE)
+}
+
+val generateWasmWasiZoneInfo = task("generateWasmWasiZoneInfo") {
```
Comment (dkhalanskyjb) [inline timezones/full/build.gradle.kts:62]:
```suggestion
val generateWasmWasiZoneInfo by tasks.registering {
```
?

File: timezones/full/build.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+                runtimeOnly(project(":kotlinx-datetime"))
+                implementation(kotlin("test"))
+            }
+        }
+    }
+}
+
+val timeTzdbInstall = task<InstallTimeTzdb>("timeTzdbInstall") { }
+
+val tzdbDownloadAndCompile = task<TzdbDownloadAndCompile>("tzdbDownloadAndCompile") {
```
Comment (dkhalanskyjb) [inline timezones/full/build.gradle.kts:54]:
```suggestion
val tzdbDownloadAndCompile by tasks.creating(TzdbDownloadAndCompile::class) {
```

File: timezones/full/build.gradle.kts [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        val wasmWasiTest by getting {
+            dependencies {
+                runtimeOnly(project(":kotlinx-datetime"))
+                implementation(kotlin("test"))
+            }
+        }
+    }
+}
+
+val timeTzdbInstall = task<InstallTimeTzdb>("timeTzdbInstall") { }
```
Comment (dkhalanskyjb) [inline timezones/full/build.gradle.kts:52]:
```suggestion
val timeTzdbInstall by tasks.creating(InstallTimeTzdb::class) { }
```

File: gradle.properties [hunk 11]
```diff
@@ -1,10 +1,13 @@
-org.gradle.jvmargs=-Xmx1G -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
+org.gradle.jvmargs=-Xmx3G -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
 org.gradle.java.installations.fromEnv=JDK_8
 
 group=org.jetbrains.kotlinx
 version=0.6.0-RC.2
 versionSuffix=SNAPSHOT
 
+timezonesMajorVersion=1.0
+tzdbVersion=2024a
```
Comment (dkhalanskyjb) [inline gradle.properties:9]: Note (we'll have to discuss this internally): SemVer doesn't allow letters in the initial components, so `1.0.2024a` is an invalid version number. We could stick to this anyway, or we could use the less natural `1.0.2024-a` (note the dash).

File: timezones/full/build.gradle.kts [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+ */
+
+import TzdbTasks.ConvertType
+import TzdbTasks.InstallTimeTzdb
+import TzdbTasks.TzdbDownloadAndCompile
+import org.jetbrains.kotlin.gradle.targets.js.dsl.ExperimentalWasmDsl
+import org.jetbrains.kotlin.gradle.targets.js.npm.NpmResolverPlugin
+
+plugins {
+    kotlin("multiplatform")
```
Comment (dkhalanskyjb) [inline timezones/full/build.gradle.kts:13]: It's a good call to have `timezonesMajorVersion` so we could describe the compatibility between `kotlinx-datetime` and the timezones package. It's easy to imagine people sticking to stable versions of the library but still upgrading the time zones.

But the Kotlin version is also part of the compatibility: there are no compatibility guarantees when someone uses the library compiled with the compiler version newer than the one used by the project itself, so to upgrade the timezone database *correctly*, you will need to upgrade Kotlin, which I think is strange.

I propose this scheme: `kotlinx-datetime-timezones-full` is compiled with the oldest compiler that works, and when some incompatibility arises between the compiler used for `kotlinx-datetime` and the one for `kotlinx-datetime-timezones-full`, only then to upgrade the compiler here, while simultaneously upgrading the `timezonesMajorVersion`.

Thoughts?

**Review by dkhalanskyjb — COMMENTED** [review]
Also, not to forget: when we decide on the versions, artifact names, etc., we'll need to update the README.

File: gradle.properties (hunk 11 above)
Comment (dkhalanskyjb) [inline gradle.properties:9]: This may be out of this PR's scope, but it's possible to automatically detect that a new version of tzdb was published: https://github.com/ThreeTen/threetenbp/blob/main/.github/workflows/tzdbupdate.yml If you feel like implementing this, it would be nice; if not, we'll just do it later ourselves.

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    public fun zoneDataByName(name: String): ByteArray
+    public fun getTimeZones(): Set<String>
+}
+
+/*
+This is internal API which is not intended to use on user-side.
+ */
+@InternalDateTimeApi
+public fun initializeTimeZonesProvider(provider: TimeZonesProvider) {
+    check(timeZonesProvider != provider) { "TimeZone database redeclaration" }
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:25]: I have no idea how to to resolve such conflicts for different flavors of timezones libraries without a special gradle plugin. We, with gradle plugin team, do not see the way for now, how to get one dependency with gradle resolution, the only way I see now is to allow redeclaration in runtime with a special "priority" level. But how to manage this level's needs a consideration.

File: core/wasmWasi/src/internal/Platform.kt [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        error("WASI call failed with $ret")
+    }
+}
+
+internal actual fun currentTime(): Instant = clockTimeGet().let { time ->
+    Instant(time / NANOS_PER_ONE, (time % NANOS_PER_ONE).toInt())
+}
+
+internal actual fun currentSystemDefaultZone(): Pair<String, TimeZoneRules?> =
+    throw UnsupportedOperationException("WASI platform does not support system timezone obtaining")
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply core/wasmWasi/src/internal/Platform.kt:42]: You would like to add it in common `TimeZone` KDoc or in make a separate KDoc for platform-specific  `currentSystemDefaultZone`?

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt (hunk 3 above)
Comment (igoriakovlev) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:15]: Well, I would prefer not suppressing anything in libraries. Only with special request from libraries team.

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt (hunk 4 above)
Comment (igoriakovlev) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:13]: Yes I think so. Do you know the way how to escape it?

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt (hunk 6 above)
Comment (igoriakovlev) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:45]: Nope, I will inline it

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt (hunk 4 above)
Comment (igoriakovlev) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:15]: It will not, because all declarations matching `.*\.internal(\..*)?` regex are suppressed.

Comment (igoriakovlev) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:15]: But now dokka NOT generates `wasm` targets (both) and generates intermediate targets `commonJs` and `pureKotlin`

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt (hunk 2 above)
Comment (igoriakovlev) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt]: You would like to add it in common `TimeZone.of` KDoc or in make a separate KDoc for platform-specific `TimeZone.of`?

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:27]: Here's a scheme:
```
kotlinx-datetime-timezones-full -> kotlinx-datetime-timezones-basic -> kotlinx-datetime
```
`A -> B` means "A depends on B". `kotlinx-datetime-timezones-full` only provides new data, the bulk of the data is in `kotlinx-datetime-timezones-basic`. Then, `kotlinx-datetime-timezones-basic` could have another thing like `initializeFullTimeZonesProvider`, which `kotlinx-datetime-timezones-full` would call. Do you see any problems with this approach?

File: core/wasmWasi/src/internal/Platform.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/wasmWasi/src/internal/Platform.kt:43]: The common one, because when writing `TimeZone.currentSystemDefault()` in common code, you should still be aware that this may fail.

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt (hunk 3 above)
Comment (dkhalanskyjb) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:17]: We prioritize user experience over how clean our internal implementation is, so yes, if it's possible to avoid extra public entry points in a robust manner, we try to do so.

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:27]: Yes, for now we are not decided yet what could `-basic` be but a first thoughts was about to remove some information from timezones (like make only `future` implementation or `+-5 years`). But not amount timezones. So the information in `timezones` then will be not additive.

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt (hunk 4 above)
Comment (dkhalanskyjb) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:15]: Yeah, you are right. It's strange that it generates something for `pureKotlin` and `commonJs`, but this PR has nothing to do with it. Thanks!

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    84,   90,  105,  102,   50,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
+    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
+    0,    0,    0,    0,    0,    0,    0,    1,    0,    0,    0,    1,    0,    0,    0,    0,
+    0,    0,    0,   84,   90,  105,  102,   50,    0,    0,    0,    0,    0,    0,    0,    0,
+    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
+    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    1,    0,    0,    0,    4,    0,
+    0,    0,    0,    0,    0,   85,   84,   67,    0,   10,   85,   84,   67,   48,   10,
+)
+
+private val utcNames get() = setOf("Universal", "Etc/Universal", "Etc/Zulu", "Etc/UCT", "Etc/UTC", "Zulu", "UCT", "UTC")
```
Comment (dkhalanskyjb) [inline core/wasmWasi/src/internal/TimeZonesInitializer.kt:44]: No-no, you don't need to do any of this: we already have special handling for `UTC` in the `TimeZone` initialization, it just needs to be present in the `availableTimeZones` set. Sorry for not being clear.

Though now that I look into it, you can even ignore that requirement, since on the JS, this rule also isn't enforced. We'll need to solve this in a centralized way.

Comment (igoriakovlev) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:44]: Ahh, so I roll it back

File: timezones/full/build.gradle.kts (hunk 12 above)
Comment (igoriakovlev) [reply timezones/full/build.gradle.kts:14]: The only thing here is that we have no possibility (in current approach) to stick a range of major versions of `timezones` to specific range of `datetime` versions in gradle. So for now it could be only documented as compatibility table.

File: gradle.properties [hunk 16]
```diff
@@ -1,10 +1,13 @@
-org.gradle.jvmargs=-Xmx1G -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
+org.gradle.jvmargs=-Xmx3G -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
```
Comment (dkhalanskyjb) [inline gradle.properties:1]: The build passes fine on my machine even without this change when I do `./gradlew tzdbDownloadAndCompile clean check`. Why is this needed?

File: buildSrc/src/main/kotlin/zoneInfosResourcesGenerator.kt [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+            zoneDataByNameBody.appendLine("    \"$name\" -> $tzDataName")
+            getTimeZonesBody.appendLine("    \"$name\",")
+        }
+    }
+
+    val content = buildString {
+        append(header)
+        appendLine()
+        appendLine()
+        appendLine("internal fun zoneDataByName(name: String): ByteArray = when(name) {")
```
Comment (dkhalanskyjb) [inline buildSrc/src/main/kotlin/zoneInfosResourcesGenerator.kt:86]: Seems extremely suboptimal to compare 600 strings just to get to `CET`. Luckily, we have the trie implementation that was recently introduced to kotlinx-datetime. I've implemented some code to do this, here:
```kotlin
private class TrieNode<T>(
    val children: MutableList<Pair<String, TrieNode<T>>> = mutableListOf(),
    var terminalValue: T? = null
)

internal fun String.toKotlinCode(): String = buildString {
    append('"')
    for (c in this@toKotlinCode) {
        when (c) {
            '"' -> append("\\\"")
            '\\' -> append("\\\\")
            '\b' -> append("\\b")
            '\t' -> append("\\t")
            '\n' -> append("\\n")
            '\r' -> append("\\r")
            else -> append(c)
        }
    }
    append('"')
}

internal fun Char.toKotlinCode(): String = when (this) {
    '\'' -> "'\\''"
    else -> "'$this'"
}

fun<V> Map<String, V>.toTrieAsKotlinCode(valueToKotlinCode: (V) -> String): String {
    val trie = TrieNode<V>()
    for ((string, v) in this) {
        var node = trie
        for (char in string) {
            val searchResult = node.children.binarySearchBy(char.toString()) { it.first }
            node = if (searchResult < 0) {
                TrieNode<V>().also { node.children.add(-searchResult - 1, char.toString() to it) }
            } else {
                node.children[searchResult].second
            }
        }
        check(node.terminalValue == null)
        node.terminalValue = v
    }
    fun reduceTrie(trie: TrieNode<V>) {
        for ((_, child) in trie.children) {
            reduceTrie(child)
        }
        val newChildren = mutableListOf<Pair<String, TrieNode<V>>>()
        for ((key, child) in trie.children) {
            if (child.terminalValue == null && child.children.size == 1) {
                val (grandChildKey, grandChild) = child.children.single()
                newChildren.add(key + grandChildKey to grandChild)
            } else {
                newChildren.add(key to child)
            }
        }
        trie.children.clear()
        trie.children.addAll(newChildren.sortedBy { it.first })
    }
    reduceTrie(trie)
    val indent = "    "
    fun StringBuilder.printTrie(trie: TrieNode<V>, inputVar: String, index: Int = 0, currentIndent: String = "    ") {
        val terminalValue = trie.terminalValue
        if (terminalValue != null) {
            append("${currentIndent}if ($inputVar.length == $index) ${valueToKotlinCode(terminalValue)} else ")
        } else {
            append(currentIndent)
        }
        if (trie.children.isEmpty()) {
            appendLine("null")
        } else {
            appendLine("when {")
            for ((key, child) in trie.children) {
                val check = if (key.length == 1) {
                    "$inputVar[$index] == ${key[0].toKotlinCode()}"
                } else {
                    "$inputVar.startsWith(${key.toKotlinCode()}, startIndex = $index)"
                }
                appendLine("${currentIndent + indent}$check -> {")
                printTrie(child, inputVar, index + key.length, currentIndent + indent + indent)
                appendLine("${currentIndent + indent}}")
            }
            appendLine("${currentIndent + indent}else -> null")
            appendLine("${currentIndent}}")
        }
    }
    return buildString { printTrie(trie, "input") }
}

// test
fun main() {
    val strings = ZoneId.getAvailableZoneIds()
    strings.associateWith { it }.toTrieAsKotlinCode { "ZoneId.of(${it.toKotlinCode()})" }.also { println(it) }
}
```

Here's a sample of what this outputs:
```kotlin
    when {
        input[0] == 'A' -> {
            when {
                input.startsWith("frica/", startIndex = 1) -> {
                    when {
                        input[7] == 'A' -> {
                            when {
                                input.startsWith("bidjan", startIndex = 8) -> {
                                    if (input.length == 14) ZoneId.of("Africa/Abidjan") else null
                                }
                                input.startsWith("ccra", startIndex = 8) -> {
                                    if (input.length == 12) ZoneId.of("Africa/Accra") else null
                                }

```

Comment (dkhalanskyjb) [reply buildSrc/src/main/kotlin/zoneInfosResourcesGenerator.kt:86]: Probably makes more sense to output the code immediately, though, and not store it as a string like I did.

File: buildSrc/src/main/kotlin/zoneInfosResourcesGenerator.kt [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+        append(zoneDataByNameBody)
+        appendLine()
+        append("    else -> throw kotlinx.datetime.IllegalTimeZoneException(\"Invalid timezone name\")")
+        appendLine()
+        append("}")
+        appendLine()
+        appendLine()
+        append("internal fun getTimeZones(): Set<String> = setOf(")
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply buildSrc/src/main/kotlin/zoneInfosResourcesGenerator.kt:94]: Well, to set memory free. I do not think this method called a lot of times and that it expected to be fast, but with `lazy` ones it called all strings never be freed.

File: buildSrc/src/main/kotlin/zoneInfosResourcesGenerator.kt (hunk 17 above)
Comment (dkhalanskyjb) [reply buildSrc/src/main/kotlin/zoneInfosResourcesGenerator.kt:84]: Oh, is that so? Cool, then there isn't much point.

File: buildSrc/src/main/kotlin/zoneInfosResourcesGenerator.kt (hunk 18 above)
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply buildSrc/src/main/kotlin/zoneInfosResourcesGenerator.kt:92]: Well, I checked it once more and see that in Wasm all loaded strings are cached in string pool. This means that once they are loaded - they are persist in memory. So yes, `lazy` is much better here.

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt (hunk 3 above)
Comment (ilya-g) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:17]: > @Suppress the warning

Note that accessing internals outside of the module is an error, not a warning.

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:27]: In any case, we'll be able to work around this somehow. This is a technical thing invisible to the users, so we can safely do anything we want here.

File: core/wasmWasi/src/internal/Platform.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/wasmWasi/src/internal/Platform.kt:43]: After an internal discussion, we decided that it makes more sense to return the UTC time zone by default. The reason is, because `TimeZone.currentSystemDefault()` is not easily mockable (see https://github.com/Kotlin/kotlinx-datetime/issues/17), sometimes, there are calls to `currentSystemDefault()` deep inside the existing code, and to have them throw exceptions now would be problematic. If/when we replace `currentSystemDefault()` with a version that supports dependency injection better, that version will throw on WASI.

This should still be documented, and it should be in the common code.

File: gradle.properties (hunk 11 above)
Comment (dkhalanskyjb) [reply gradle.properties:9]: In the end, this isn't a SemVer anyway, so dashes aren't needed, `2024a` is ok.

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt (hunk 3 above)
Comment (dkhalanskyjb) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:17]: Ok, in that case, why not `@Suppress` the error?

File: settings.gradle.kts [hunk 19]
```diff
@@ -16,6 +16,8 @@ rootProject.name = "Kotlin-DateTime-library"
 
 include(":core")
 project(":core").name = "kotlinx-datetime"
+include(":timezones/full")
+project(":timezones/full").name = "kotlinx-datetime-timezones-full"
```
Comment (dkhalanskyjb) [inline settings.gradle.kts:20]: The artifact name we decided on is `kotlinx-datetime-zoneinfo`. At least for now, without `full`.

Comment (igoriakovlev) [reply settings.gradle.kts:20]: And if we decide to make an intermediate timezones db?

Comment (dkhalanskyjb) [reply settings.gradle.kts:20]: When we decide to do it, we'll know from the reported use cases what exactly people want, and then we'll choose the names appropriately.

File: core/wasmWasi/src/internal/Platform.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/wasmWasi/src/internal/Platform.kt:43]: Yes.

File: settings.gradle.kts (hunk 19 above)
Comment (dkhalanskyjb) [reply settings.gradle.kts]: `-zoneinfo`, not `-timezones`.

**Comment by igoriakovlev** [comment]
I have rebased onto the master and merged all the AMEND commits by @dkhalanskyjb request.

**Review by dkhalanskyjb — APPROVED** [review]

File: core/wasmWasi/src/internal/TimeZonesInitializer.kt (hunk 3 above)
Comment (ilya-g) [reply core/wasmWasi/src/internal/TimeZonesInitializer.kt:17]: This is not advisable by the compiler team and can stop working in K2

**Review by ilya-g — CHANGES_REQUESTED** [review]

File: timezones/full/build.gradle.kts [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+val tzdbDownloadAndCompile by tasks.creating(TzdbDownloadAndCompile::class) {
+    dependsOn(timeTzdbInstall)
+    timeTzdbDirectory.set(timeTzdbInstall.outputDirectory)
+    outputDirectory.set(tzdbDirectory)
+    ianaVersion.set(tzdbVersion)
+    convertType.set(ConvertType.LARGE)
+}
+
+val generateWasmWasiZoneInfo by tasks.registering {
+    inputs.dir(tzdbDirectory)
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply timezones/full/build.gradle.kts:71]: As we agreed with @dkhalanskyjb it is going to be committed.

File: buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+
+        val outputDir = project.layout.buildDirectory.zip(typePostfix) { build, type -> build.dir("tzdbCompiled$type") }
+
+        outputDirectory.convention(outputDir)
+        convertType.convention(ConvertType.LARGE)
+        executable = ""
+    }
+
+    override fun exec() {
+        executable = File(timeTzdbDirectory.get().asFile, "node_modules/.bin/tzc").path
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt:70]: You sure we need to support this? The result of this is going to be committed so this called only when someone wants to update tzdb.

File: buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+    @get:Optional
+    abstract val outputDirectory: DirectoryProperty
+
+    @get:Input
+    @get:Optional
+    abstract val version: Property<String>
+
+    init {
+        outputDirectory.convention(project.layout.buildDirectory.dir("time-tzdb"))
+        executable = "npm"
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt:24]: > Would it be possible to use npm downloaded by gradle? 
As far as I know there is no good way to download npm package with gradle.

> Or at least allow specifying a full path using a gradle property?
full path to what? to the `npm`

File: core/build.gradle.kts [hunk 23]
```diff
@@ -221,14 +225,34 @@ kotlin {
             dependsOn(commonJsTest)
         }
 
-        val nativeMain by getting {
+        val pureKotlinMain by creating {
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply core/build.gradle.kts]: done

File: timezones/full/build.gradle.kts [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+    outputs.dir(convertedKtFilesDir)
+    doLast {
+        generateZoneInfosResources(tzdbDirectory, convertedKtFilesDir, tzdbVersion)
+    }
+}
+
+tasks.getByName("compileKotlinWasmWasi") {
+    dependsOn(generateWasmWasiZoneInfo)
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply timezones/full/build.gradle.kts]: done

File: buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt (hunk 22 above)
[the remark this answers is not in the document]
Comment (ilya-g) [reply buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt:24]: > As far as I know there is no good way to download npm package with gradle.

Isn't it what `node` gradle plugin does? https://github.com/node-gradle/gradle-node-plugin

File: buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt (hunk 21 above)
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt:70]: Do we really need support this?

File: buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt (hunk 22 above)
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt]: Well, with this plugin we could make something like this:
https://github.com/Kotlin/kotlinx-datetime/pull/366/commits/6832cde6562d6204225a7b9933a8caefe1ae465e

[the remark this answers is not in the document]
Comment (ilya-g) [reply buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt]: Yes, that's better. Perhaps we can replace the second task with NpxTask, though I'm not sure.

[the remark this answers is not in the document]
Comment (igoriakovlev) [reply buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt]: I tried it [here](https://github.com/Kotlin/kotlinx-datetime/commit/8c2e9b1e71eed7abe8b7d1aeabcfe71589b2cf89)
I could not change a directory for npm project (`node.nodeProjectDir` did't work) so now `node_modules` located in the project dir

File: timezones/full/build.gradle.kts [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+
+val timeTzdbInstall by tasks.creating(NpmTask::class) {
+    args.addAll(
+        "install",
+        "@tubular/time-tzdb",
+    )
+}
+
+val tzdbDownloadAndCompile by tasks.creating(NpxTask::class) {
+    tzdbDirectory.mkdirs()
```
Comment (ilya-g) [inline timezones/full/build.gradle.kts:32]: Please move it to the execution phase, e.g. by wrapping in `doFirst { }`

**Review by ilya-g — CHANGES_REQUESTED** [review]

File: timezones/full/build.gradle.kts [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+
+        val commonTest by getting {
+            dependencies {
+                runtimeOnly(project(":kotlinx-datetime"))
+                implementation(kotlin("test"))
+            }
+        }
+
+        val wasmWasiMain by getting {
+            languageSettings.optIn("kotlinx.datetime.internal.InternalDateTimeApi")
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply timezones/full/build.gradle.kts:86]: No, it does not work if you apply it on common code, only when you compile target module.

File: timezones/full/build.gradle.kts (hunk 25 above)
Comment (igoriakovlev) [reply timezones/full/build.gradle.kts]: Done

File: .gitignore [hunk 27]
```diff
@@ -6,3 +6,7 @@ target
 build
 /local.properties
 benchmarks.jar
+node_modules
+package.json
+timezones.json
+package-lock.json
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply .gitignore]: Done

File: timezones/full/build.gradle.kts [hunk 28]
```diff
@@ -0,0 +1,84 @@
...
+    dependsOn(timeTzdbInstall)
+    command.set("@tubular/time-tzdb")
+    args.addAll("-b", "-o", "--large")
+    if (tzdbVersion.isNotEmpty()) {
+        args.addAll("-u", tzdbVersion)
+    }
+    args.add(tzdbDirectory.toString())
+}
+
+val generateWasmWasiZoneInfo by tasks.registering {
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply timezones/full/build.gradle.kts]: Done

File: buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt (hunk 22 above)
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply buildSrc/src/main/kotlin/TzdbTasks/TzdbTasks.kt]: Seems works indeed, done

File: timezones/full/build.gradle.kts (hunk 26 above)
[the remark this answers is not in the document]
Comment (ilya-g) [reply timezones/full/build.gradle.kts:88]: Strange, but ok. We'll move it to sourceSets.all { } if we'll be commonizing this artifact further.

**Review by ilya-g — APPROVED** [review]
We'll wait for 0.6.1 to be released from master before merging this

**Comment by igoriakovlev** [comment]
I sqashed commits so PR is ready to merge. All tests passed. Thank you a lot for the review!

**Comment by igoriakovlev** [comment]
> If this is good to go, could you rebase it on top of `master` so that we can merge it?

It is done.

**Comment by dkhalanskyjb** [comment]
Great job, thanks!
